pkg/sdk/errors.go

```diff
@@ -48,3 +47,230 @@ func decodeDriverError(err error) error {
+// TODO this approach to predefined errors, this is lazy evaluated error, so the only filename + line number
+// TODO if we're going with this approach, change other methods to handle SDKPredefinedError type
+func NewPredefinedError2(message string) SDKPredefinedError {
+	err := newSDKError(message, 2)
+	return func() error {
+		return err
+	}
+}
+
+func NewErrorOneOf(structPointer any, fields ...any) error {
+	structure := reflect.ValueOf(structPointer).Elem()
+	var fieldsBuilder strings.Builder
+
+	for fieldIndex := range fields {
+		fieldMeta := structure.Type().Field(fieldIndex)
+		fieldValue := structure.Field(fieldIndex)
+		fieldPointer := reflect.NewAt(fieldMeta.Type, unsafe.Pointer(fieldValue.UnsafeAddr()))
+		if fieldMeta.Type.Kind() == reflect.Pointer {
+			fieldPointer = fieldPointer.Elem()
+		}
+
+		fieldsBuilder.WriteString(fmt.Sprintf("%s %s(%v)", fieldMeta.Name, fieldMeta.Type, fieldPointer.Elem().Interface()))
+
+		if fieldIndex != len(fields)-1 {
+			fieldsBuilder.WriteString(", ")
+		}
+	}
+
+	return newSDKError(
+		fmt.Sprintf(
+			"fields of struct %s [%s] are incompatible and shouldn't be set at the same time",
+			structure.Type().Name(),
+			fieldsBuilder.String(),
+		),
+		2,
+	)
+}
+
+func NewErrorNotSet(structPointer any, fields ...any) error {
+	structure := reflect.ValueOf(structPointer).Elem()
+	var fieldsBuilder strings.Builder
+
+	for fieldIndex := range fields {
+		fieldMeta := structure.Type().Field(fieldIndex)
+		fieldsBuilder.WriteString(fmt.Sprintf("%s %s", fieldMeta.Name, fieldMeta.Type))
+
+		if fieldIndex != len(fields)-1 {
+			fieldsBuilder.WriteString(", ")
+		}
+	}
+
+	return newSDKError(
+		fmt.Sprintf(
+			"fields of struct %s: [%s] are required and should be set",
+			structure.Type().Name(),
+			fieldsBuilder.String(),
+		),
+		2,
+	)
+}
+
+// WrapErrors wraps errs with wrapper error.
+// When wrapper is any other type than SDKError than it'll create new SDKError with wrapper's error message
+// and in every case errs will be added to its internal list of errors.
+// TODO We can force to use SDKError as wrapper
+func WrapErrors(wrapper error, errs ...error) error {
+	if err, ok := wrapper.(*SDKError); ok { //nolint:all
+		err.nestedErrors = append(err.nestedErrors, errs...)
+		return wrapper
+	} else {
+		return newSDKError(wrapper.Error(), 2, errs...)
+	}
+}
+
+func newSDKError(message string, skip int, nested ...error) error {
+	line, filename := getCallerInfo(skip)
+	return &SDKError{
+		file:         filename,
+		line:         line,
+		message:      message,
+		nestedErrors: nested,
+	}
+}
+
+func getCallerInfo(skip int) (int, string) {
+	_, file, line, _ := runtime.Caller(skip + 1)
+	fileSplit := strings.Split(file, "/")
+	var filename string
+	if len(fileSplit) > 1 {
+		filename = fileSplit[len(fileSplit)-1]
+	} else {
+		filename = fileSplit[0]
+	}
+	return line, filename
+}
+
+func writeTree(e error, builder *strings.Builder, indent int, fileInfo bool) {
```

Comment:
This method needs to be thoroughly tested.

Reply:
Added tests